rfd/00TBD-webui-paginate-filter.md

```diff
@@ -0,0 +1,343 @@
+---
+authors: Lisa Kim (lisa@goteleport.com)
+state: draft
+---
+
+# RFD TBD - WebUI server-side paginating and filtering
+
+## What
+
+Provide server-side pagination and filtering capabilities for the web UI for select resources: nodes, apps, dbs, kubes, and desktops.
+
+## Why
+
+Currently, the web api calls upon non paginated endpoints that retrieves an entire list of resources which for higher ranges ~20k+, results in a ~30+ second load time for the UI. With pagination support, we can control the limit of resources to fetch, increasing load speed, and user experience. Filtering will then need to be done on the server-side to apply it to the entire list of resources.
+
+## Details
+
+### Pagination
+
+There is already pagination support for resources: nodes, apps, dbs. Kubes is [pending](https://github.com/gravitational/teleport/pull/9096#issuecomment-989291135)
+
+### Filter
+
+There will be filters that will be applied in order of precedence: RBAC > label > search > sort. RBAC will not be discussed since this check existed before. 
+
+### Filter: User Label Query Language
+
+The query language for labels will be modeled after [kubernetes labels](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/). Kubernetes uses `set based` filtering which matches a label `key` against a set of `values` as shown in examples below.
+
+The language will be the same across the web ui and cli tools.
+
+Supported operators:
```

Comment:
When it comes to the syntax, we would achieve better results if we start with the use cases. Let's also not forget that we already have a syntax (predicate) used in our RBAC, so we have to be careful not to introduce another, completely different syntax. As a bonus, we already have it all implemented and tested with many features and operators, like `equals, contains, regexp` and others.

You can see it here:

https://goteleport.com/docs/access-controls/guides/impersonation/#dynamic-impersonation
https://github.com/gravitational/teleport/blob/master/lib/services/parser_test.go

Here is what predicate can do:

1. most common one, filter by label:

```go
resource.metadata.labels["env"] == "prod"
```

2. filter by multiple labels and resource name 

```go
resource.metadata.labels["env"] == "prod" && resources.metadata.labels["staging"] != "test" || resource.metadata.name != "jenkins"
```

Reply:
@klizhentas my [first iteration](https://github.com/gravitational/teleport/blob/05bd127739b2ee7e29afaaa1dcb487cf8370f754/rfd/00TBD-webui-paginate-filter.md#filter-label--search) of this RFD (ignoring the search values, idk why I thought it was good idea to combine it then) suggested the use of predicate library, but under the hood. It still suggested a user facing query language (b/c I thought we may want a more simpler language for the user), that then got converted into predicate language. 

But what you are suggesting is to have user use the predicate language as is, like we already do for our roles. I am okay with this. It would break compatibility with how `tsh` currently allows users to query for labels which is:
```
tsh ls key=value,key2=value2
```
Or we could keep tsh simple label query as is, and add a long flag `--label` that allows you to use predicate language:
```
tsh ls --label=`equals("env", "prod") && !equals("staging", "test")`
```

This is what you are suggesting right?

Reply:
`tsh ls --query` would be more appropriate, as it's just not for labels. Yes, I suggest to keep predicate as the only language for backend and frontend.

Reply:
@kimlisa keep in mind, that you can still make a very short expression in predicate, these are all valid predicate expressions:

```go
// if key is a variable
key
```

```go
// you can define labels dict/struct and access it's properties with .
labels.key == `hello`
```

```go
name != "Jenkins"
```